Refactor Stream Modify and Deallocate to perform validation checks in the SDK cluster code.

pidarped opened the pull request:
#### Summary
Move the required error checks for Audio/Video/Snapshot stream modification and deallocation from the delegate to the common SDK cluster.


#### Related issues

#### Testing
* All AVSM test scripts pass when run against chip-camera-app.
 
#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

pidarped committed:
Refactor Stream Modify and Deallocate to perform validation checks
in the SDK cluster code.

- Move the required error checks for Audio/Video/Snapshot stream
  modification and deallocation from the delegate to the common SDK
  cluster.

pidarped committed:
Add template function for removing some logic duplication.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -2238,6 +2260,90 @@ bool CameraAVStreamMgmtServer::ValidateSnapshotStreamId(const DataModel::Nullabl
+    }
+
+    return true;
+}
```
Incorporated in latest commit.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -2238,6 +2260,90 @@ bool CameraAVStreamMgmtServer::ValidateSnapshotStreamId(const DataModel::Nullabl
+    return true;
+}
+
+bool CameraAVStreamMgmtServer::ValidateSnapshotStreamForModifyOrDeallocate(const uint16_t snapshotStreamID, HandlerContext & ctx)
```
StreamUsage is not part of SnapshotStreams at the moment.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -600,6 +600,42 @@ class CameraAVStreamMgmtServer : public CommandHandlerInterface, public Attribut
+        {
+            ChipLogError(Zcl, "CameraAVStreamMgmt[ep=%d]: %s stream with ID: %u still in use", mEndpointId, streamTypeName,
+                         streamID);
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::InvalidInState);
```
Where does the spec say you can't modify a stream with nonzero refcount?

pidarped replied:
Yes, the ref count checks are only performed on Stream Deallocation commands.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -600,6 +600,42 @@ class CameraAVStreamMgmtServer : public CommandHandlerInterface, public Attribut
+        using StreamType = typename StreamContainer::value_type;
+        if constexpr (std::is_same_v<StreamType, VideoStreamStruct> || std::is_same_v<StreamType, AudioStreamStruct>)
+        {
+            if (it->streamUsage == Globals::StreamUsageEnum::kInternal)
```
Where is this check in the spec for Modify?  I don't see it there.

Or at least it's equally there or not for both snapshot and audio/video streams (unlike Deallocate, where it's there for audio/video but not snapshot).

Or is the point that snapshot streams are never internal?  I see nothing in the spec that guarantees that.

pidarped replied:
Apologies. There is a bit of flux here. It is there in the current Spec for Audio/Video StreamDeallocate. And it is getting added for `VideoStreamModify` in https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/12117. Expected the Spec PR to have/be merged soon.
SnapshotStreams, OTOH, can only be Internal if they are being derived from a VideoStream that is Internal. In that scenario, SnapshotStreamModify would be disallowed. This bug was found recently and filed here: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/12126.
Let's hold off on the review until the Spec side is in then as I would need to update the SnapshotStreamModify as well.

bzbarsky-apple replied:
Ah, the usual "we are implementing an in-flight PR but not linking to it" thing?  Alright then.

pidarped committed:
Update validation in template function for Stream Modify

-Stream deallocations only check against ref counting
-SnapshotStreamModify needs to be checked against
 requiresHardwareEncoder.

pidarped commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.cpp:
```diff
@@ -2238,6 +2260,24 @@ bool CameraAVStreamMgmtServer::ValidateSnapshotStreamId(const DataModel::Nullabl
 
+bool CameraAVStreamMgmtServer::ValidateVideoStreamForModifyOrDeallocate(const uint16_t videoStreamID, HandlerContext & ctx)
+{
+    return ValidateStreamForModifyOrDeallocateImpl(mAllocatedVideoStreams, videoStreamID, ctx, "Video",
```
The ref counts are updated in both the allocated streams in CameraAVStreamMgmtServer and the available streams in the CameraDevice.

pidarped committed:
Define an enum for the Stream types and pass to the template for
validation instead of passing string constants.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -617,6 +639,66 @@ class CameraAVStreamMgmtServer : public CommandHandlerInterface, public Attribut
+            return false;
+        }
+
+        using StreamValueType = typename StreamContainer::value_type;
```
Since you know this, I do wonder why you need a passed-in StreamType, but ok.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h:
```diff
@@ -617,6 +639,66 @@ class CameraAVStreamMgmtServer : public CommandHandlerInterface, public Attribut
+                    std::find_if(mSnapshotCapabilitiesList.begin(), mSnapshotCapabilitiesList.end(), [&](const auto & capability) {
+                        return capability.imageCodec == it->imageCodec && capability.maxFrameRate > it->frameRate;
+                    });
+                if (sn_capab_it != mSnapshotCapabilitiesList.end() && sn_capab_it->requiresHardwareEncoder.HasValue() &&
+                    !sn_capab_it->requiresHardwareEncoder.Value())
```
Could have been simpler as `sn_capab_it->requiresHardwareEncoder.ValueOr(false)`, right?
